### PR TITLE
UNR-1857  Unresolved RPC parameters can cause permanent queue blocking

### PR DESCRIPTION
**Pull request by aleximprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Now RPCs that have been queued for 1 second are processed with unresolved parameters.

#### Release note
Updated CHANGELOG.md.

#### Tests
Verified that client/server doesn't get stuck because of queued RPCs (in local deployment).

#### Documentation
Updated CHANGELOG.md.

#### Primary reviewers
@m-samiec @improbable-valentyn

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 1]
```diff
@@ -69,5 +70,21 @@ bool FRPCContainer::ObjectHasRPCsQueuedOfType(const Worker_EntityId& EntityId, E
...
-	return FunctionToApply.Execute(Params);
+	if(FunctionToApply.Execute(Params))
+	{
+		return true;
+	}
+	else
+	{
+		FTimespan TimeDiff = FDateTime::Now() - Params.Timestamp;
+		bool bDropRPC = TimeDiff.GetSeconds() > SECONDS_TO_DROP_RPC;
+		if (bDropRPC)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]: just return bDropRPC?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 2]
```diff
@@ -67,7 +69,25 @@ bool FRPCContainer::ObjectHasRPCsQueuedOfType(const Worker_EntityId& EntityId, E
 	return false;
 }
 
-bool FRPCContainer::ApplyFunction(const FProcessRPCDelegate& FunctionToApply, const FPendingRPCParams& Params)
+bool FRPCContainer::ApplyFunction(const FProcessRPCDelegate& FunctionToApply, FPendingRPCParams& Params)
 {
-	return FunctionToApply.Execute(Params);
+	if(FunctionToApply.Execute(Params))
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]:
```suggestion
	if (FunctionToApply.Execute(Params))
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -1494,7 +1494,8 @@ bool USpatialReceiver::ApplyRPC(UObject* TargetObject, UFunction* Function, cons
 	TSharedPtr<FRepLayout> RepLayout = NetDriver->GetFunctionRepLayout(Function);
 	RepLayout_ReceivePropertiesForRPC(*RepLayout, PayloadReader, Parms);
 
-	if (UnresolvedRefs.Num() == 0)
+	if ((UnresolvedRefs.Num() == 0)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Nit: prefer this on one line.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h [hunk 4]
```diff
@@ -13,13 +13,16 @@ using FPendingRPCParamsPtr = TUniquePtr<FPendingRPCParams>;
 DECLARE_DELEGATE_RetVal_OneParam(bool, FProcessRPCDelegate, const FPendingRPCParams&)
 
 struct FPendingRPCParams
-{
+{
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]:
```suggestion
{
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h [hunk 5]
```diff
@@ -35,7 +38,8 @@ class FRPCContainer
 	using RPCContainerType = TMap<ESchemaComponentType, FRPCMap>;
 
 	void ProcessRPCs(const FProcessRPCDelegate& FunctionToApply, FArrayOfParams& RPCList);
-	static bool ApplyFunction(const FProcessRPCDelegate& FunctionToApply, const FPendingRPCParams& Params);
+	static bool ApplyFunction(const FProcessRPCDelegate& FunctionToApply, FPendingRPCParams& Params);
 
 	RPCContainerType QueuedRPCs;
+	static const int32 SECONDS_TO_DROP_RPC = 4;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: Could you make this a float and put it in the `SpatialGDKSettings`?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 6]
```diff
@@ -67,7 +69,25 @@ bool FRPCContainer::ObjectHasRPCsQueuedOfType(const Worker_EntityId& EntityId, E
...
+	{
+		return true;
+	}
+	else
+	{
+		FTimespan TimeDiff = FDateTime::Now() - Params.Timestamp;
+		bool bProcessAnywayOrDropRPC = TimeDiff.GetSeconds() > SECONDS_TO_DROP_RPC;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]:
```suggestion
		bool bProcessAnywayOrDropRPC = TimeDiff.GetTotalSeconds() > SECONDS_TO_DROP_RPC;
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: I agree. I also feel the default should be a lot shorter.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 7]
```diff
@@ -67,7 +69,25 @@ bool FRPCContainer::ObjectHasRPCsQueuedOfType(const Worker_EntityId& EntityId, E
...
 
-bool FRPCContainer::ApplyFunction(const FProcessRPCDelegate& FunctionToApply, const FPendingRPCParams& Params)
+bool FRPCContainer::ApplyFunction(const FProcessRPCDelegate& FunctionToApply, FPendingRPCParams& Params)
 {
-	return FunctionToApply.Execute(Params);
+	if(FunctionToApply.Execute(Params))
+	{
+		return true;
+	}
+	else
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]: The else part should be only executed on the receiving side.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: What value you think suits best @m-samiec ?

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: I'll put 3 seconds for now

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: My opinion would be 1 second. I think if we have logs for when this is occurring, we can run a full playtest and evaluate the logs

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp (hunk 6 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]: Applied.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: Changed to 1 second, will add logs.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp (hunk 2 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]: Resolving as outdated.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp (hunk 1 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp]: Resolving as outdated.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 3 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Updated.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: Moved to settings, added logs.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h [hunk 8]
```diff
@@ -20,6 +20,8 @@ struct FPendingRPCParams
 	int ReliableRPCIndex;
 	FUnrealObjectRef ObjectRef;
 	SpatialGDK::RPCPayload Payload;
+
+	FDateTime Timestamp;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: nit: ReceivedTimestamp ?

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 9]
```diff
@@ -123,6 +123,10 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
 	UPROPERTY(config, meta = (ConfigRestartRequired = false))
 	bool bCheckRPCOrder;
 
+	/** Seconds to wait before executing an RPC substituting null values for unresoled parameters */
+	UPROPERTY(config, meta = (ConfigRestartRequired = false))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: I think we should include this in the visible GDK settings under the category "Replication"

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 10]
```diff
@@ -123,6 +123,10 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
...
+	float SecondsToProcessRPCWithUnresolvedRefs;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: Bit of a mouthful - but I think I prefer `WaitTimeBeforeProcessingReceivedRPCWithUnresolveRefs`

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 11]
```diff
@@ -123,6 +123,10 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
 	UPROPERTY(config, meta = (ConfigRestartRequired = false))
 	bool bCheckRPCOrder;
 
+	/** Seconds to wait before executing an RPC substituting null values for unresoled parameters */
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]:
```suggestion
	/** Seconds to wait before executing a received RPC substituting nullptr for unresolved UObjects*/
```

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h (hunk 10 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: It's already a mouthful, so I'll change as you suggest :)

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 8 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: QueuedTimestamp?

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h (hunk 10 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: Updated.

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h (hunk 9 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: Added.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 8 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: Sure

File: CHANGELOG.md [hunk 12]
```diff
@@ -5,6 +5,7 @@ The format of this Changelog is based on [Keep a Changelog](https://keepachangel
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased-`x.y.z`] - 2019-xx-xx
+- RPCs that have been queued for execution for more than 1 second (default value in SpatialGDKSettings SecondsToProcessRPCWithUnresolvedRefs) are executed with unresolved parameters.
```
Comment (m-samiec) [inline CHANGELOG.md]:
```suggestion
- RPCs that have been queued for execution for more than 1 second (default value in SpatialGDKSettings WaitTimeBeforeProcessingReceivedRPC_WithUnresolvedRefs) are executed with unresolved parameters.
```

Comment (aleximprobable) [reply CHANGELOG.md]: Good catch!

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 8 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h]: Renamed.

File: CHANGELOG.md [hunk 13]
```diff
@@ -5,6 +5,7 @@ The format of this Changelog is based on [Keep a Changelog](https://keepachangel
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased-`x.y.z`] - 2019-xx-xx
+- RPCs that have been queued for execution for more than 1 second (default value in SpatialGDKSettings WaitTimeBeforeProcessingReceivedRPC_WithUnresolvedRefs) are executed with unresolved parameters.
```
Comment (improbable-valentyn) [inline CHANGELOG.md]:
```suggestion
- RPCs that have been queued for execution for more than 1 second (default value in SpatialGDKSettings QueuedIncomingRPCWaitTime) are executed with unresolved parameters.
```

**Pull request merged**